### PR TITLE
TestUtils & sendEmails

### PR DESCRIPTION
**Pull request by Yizhuzhu**
Test DrugMapUtils & FirebasePathUtils

### CODE REVIEW

File: web/yo/app/scripts/services/user.js [hunk 1]
```diff
@@ -11,7 +11,7 @@
  * gmail.com will be used as standard email format, googlemail address will be converted to gmail
  */
 angular.module('oncokbApp')
-    .service('user', function user($routeParams, $q, $firebaseAuth, $firebaseObject, $rootScope, mainUtils, firebaseConnector) {
+    .service('user', function user($routeParams, $q, $firebaseAuth, $firebaseObject, $rootScope, mainUtils, firebaseConnector, dialogs) {
```
Comment (zhx828) [inline web/yo/app/scripts/services/user.js]: No change needed?

File: web/yo/app/scripts/services/drugmaputils.spec.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            "description" : "",
+            "drugName" : "Ponatinib",
+            "ncitCode" : "C95777",
+            "ncitName" : "Ponatinib",
+            "synonyms" : [ "PONATINIB", "AP-24534", "AP24534", "Benzamide, 3-(2-Imidazo(1,2-B)Pyridazin-3-Ylethynyl)-4-Methyl-N-(4-((4-Methyl-1- Piperazinyl)Methyl)-3-(Trifluoromethyl)Phenyl)" ],
+            "uuid" : "122b8921-3cd6-4b57-bdc8-5d39cc5465a1"
+        }
+    };
+    it('therapyStrToArr should change therapyName from String to Array', function () {
+        expect(drugMapUtilsFactory.therapyStrToArr('Encorafenib + Pembrolizumab, MDM2 Antagonist RO5045337, MDM2 Antagonist RO5045337 + Nilotinib, Ponatinib')).toEqual([['Encorafenib', 'Pembrolizumab'], ['MDM2 Antagonist RO5045337'], ['MDM2 Antagonist RO5045337', 'Nilotinib'], ['Ponatinib']]);
```
Comment (zhx828) [inline web/yo/app/scripts/services/drugmaputils.spec.js]: Could you also do some extreme tests, like  if the input is null, undefined or empt string? You could also try to test some format cases. For instance `Encorafenib +Pembrolizumab`, `Encorafenib+ Pembrolizumab`, ` Encorafenib+Pembrolizumab ` is supposed to be `([['Encorafenib', 'Pembrolizumab']]`

File: web/yo/app/scripts/services/user.js (hunk 1 above)
Comment (Yizhuzhu) [reply web/yo/app/scripts/services/user.js]: dialogs has been added to master branch. I have no idea why there is difference here. dialogs is needed.

File: web/yo/app/scripts/services/drugmaputils.spec.js (hunk 2 above)
Comment (Yizhuzhu) [reply web/yo/app/scripts/services/drugmaputils.spec.js]: For therapyStrToArray function, the input can't be null or undefined. It has been checked before triggering this function. And removing duplicates is not needed and supported here. For other functions, some extreme tests have been included.

Comment (zhx828) [reply web/yo/app/scripts/services/drugmaputils.spec.js]: You don't know where it's going to call this function, so the input is not always guaranteed. If we have typescript, these do not need to be tested. But for now, let's have some test coverages.

File: web/yo/app/scripts/services/drugmaputils.js [hunk 3]
```diff
@@ -11,9 +11,11 @@ angular.module('oncokbApp')
     .factory('drugMapUtils', function(OncoKB, _, $q, DatabaseConnector, $rootScope, ReviewResource, S, UUIDjs, $routeParams, firebaseConnector, firebasePathUtils) {
 
         function therapyStrToArr(key){
-            return key.split(",").map(function(element){
-                return element.trim().split(" + ")
-            });
+            if(key){
```
Comment (zhx828) [inline web/yo/app/scripts/services/drugmaputils.js]: This will return undefined for condition not passed, is this expected?

File: web/yo/app/scripts/services/mainutils.js [hunk 4]
```diff
@@ -285,18 +285,29 @@ angular.module('oncokbApp')
...
-                param,
-                function(result) {
-                    deferred.resolve(result);
-                },
-                function(result) {
-                    deferred.reject(result);
-                }
-            );
-            return deferred.promise;
+            if(content){
```
Comment (zhx828) [inline web/yo/app/scripts/services/mainutils.js]: Also do a sendTo check.

File: web/yo/app/scripts/services/drugmaputils.js (hunk 3 above)
Comment (Yizhuzhu) [reply web/yo/app/scripts/services/drugmaputils.js]: Yes. Generally, it should always be passed, because the key is checked before calling this function. I set this condition just for testing. If not, split will lead to error.

File: web/yo/app/scripts/services/mainutils.js (hunk 4 above)
Comment (Yizhuzhu) [reply web/yo/app/scripts/services/mainutils.js]: I think check might not be needed here, because there is a promise. If sendTo is null, the return should be reject. If I did check here, there might not be reminder.

Comment (zhx828) [reply web/yo/app/scripts/services/mainutils.js]: You should handle the empty/undefined case and reject the promise manually. The promise you referenced is the one actually sending to backend for the service. But the whole purpose of checking is to avoid that.

File: web/yo/app/scripts/services/drugmaputils.js (hunk 3 above)
Comment (zhx828) [reply web/yo/app/scripts/services/drugmaputils.js]: Should it return undefined or return empty string.
